### ISSUE TITLE
questions about batch norm auto-upgrade.

### ISSUE THREAD

**Issue by D-X-Y** [BVLC/caffe#5120]
### In Upgrade_proto the param of bn layer is clear 
In upgrade_proto.cpp,  the param of bn layer is cleared but I want to set param::name and it's cleared.
eg. two bn layers share parameters...
So 
```
if (net_param.layer(i).type() == "BatchNorm"
        && net_param.layer(i).param_size() == 3) { 
      return true;
}   
```
seems wrong ...

**Comment by shaibagon** [comment]
@D-X-Y  missed that. thanks for spotting. can these two issues be merged?

**Comment by D-X-Y** [comment]
@shaibagon Of course! But how to merge these two issues ? I'm not good at github T.T

**Comment by shelhamer** [comment]
Closing as #5171 has more detail on this same issue. Thanks for the report.

**Comment by BrettLL** [comment]
@D-X-Y I have the problem that why should set lr_mult: 0 three times in BatchNorm layer?? Is it needed in both train and test phase?

**Comment by D-X-Y** [comment]
@BrettLL Yes, it is need in both train and test phase. Because the parameters in BatchNorm layer is updated in the forward function and the diff_ is not used for parameters in BatchNorm layer.
If you do not set lr_mult=0, the solver will change the parameters in BatchNorm layer, which is incorrect.

**Comment by BrettLL** [comment]
@D-X-Y  1. Is the parameters "mean, variance, moving average factor"?? 2.They has no diff_ and if we don't set lr_mult=0, how can it update??  By the way, the code in your git: https://github.com/D-X-Y/caffe-reid/blob/master/models/market1501/res50/train.proto, should we set the use_global_stats false in train phase in BatchNorm??

**Comment by D-X-Y** [comment]
@BrettLL I think they has the diff_ variable but not use, so the values in diff_ are meaningless.
When finetuning, the use_global_stats should be set to true.

**Comment by BrettLL** [comment]
@D-X-Y when set use_global_stats true, it means the parameters in BatchNormLayer will not be changed. Is it suitable for different tasks??  I think the reason why set it true in finetuning is lacking of data, If we have big data, it should be set false in finetuning progress, how do you think it?

**Comment by D-X-Y** [comment]
As I know, fix BN layer is suitable for vaious tasks, such as object detection, segamentation, action recogination, person re-id.
It may be complicated to explain why fix, the intuitive reseason is the data size.
I think you can try both fix and not fix.
Further, if you have a much larger data than ImageNet and the data distribution is quite different , why not directly train on your own data from scrash.

**Comment by BrettLL** [comment]
@D-X-Y  Yes, I think you are right. I am trying both fix and not fix on person re-id task, It seems have little difference between them~ If I have larger data, I would prefer not to fix or train from scratch.  Thanks for your reply~~
